pydatarecognition/runcontrol.py

```diff
@@ -0,0 +1,298 @@
+
+    def _validate(self, key, value):
+        """Validates - and possibly converts - a value based on its key and the current
+        validators.
+        """
+        validators = self._validators
+        if key in validators:
+            validator, convertor = validators[key]
+        else:
+            for vld in validators:
+                if isinstance(vld, str):
+                    continue
+                m = vld.match(key)
+                if m is not None:
+                    validator, convertor = validators[vld]
+            else:
+                validator, convertor = always_true, noop
+        return value if validator(value) else convertor(value)
+
+
+def flatten(iterable):
+    """Generator which returns flattened version of nested sequences."""
+    for el in iterable:
+        if isinstance(el, basestring):
+            yield el
+        elif isinstance(el, Iterable):
+            for subel in flatten(el):
+                yield subel
+        else:
+            yield el
+
+
+#
+# Memoization
+#
+
+
+def ishashable(x):
+    """Tests if a value is hashable."""
+    if isinstance(x, Hashable):
+        if isinstance(x, basestring):
+            return True
+        elif isinstance(x, Iterable):
+            return all(map(ishashable, x))
+        else:
+            return True
+    else:
+        return False
+
+
+DEFAULT_RC = RunControl(
+    _validators=DEFAULT_VALIDATORS,
+    builddir="_build",
+    mongodbpath=property(lambda self: os.path.join(self.builddir, "_dbpath")),
+    user_config=os.path.expanduser("~/.config/regolith/user.json"),
+    force=False,
+    database=None
+)
+
+
+def load_json_rcfile(fname):
+    """Loads a JSON run control file."""
+    with open(fname, "r", encoding='utf-8') as f:
+        rc = json.load(f)
+    return rc
+
+
+def load_rcfile(fname):
+    """Loads a run control file."""
+    base, ext = os.path.splitext(fname)
+    if ext == ".json":
+        rc = load_json_rcfile(fname)
+    else:
+        raise RuntimeError(
+            "could not detemine run control file type from extension."
+        )
+    return rc
+
+
+def filter_databases(rc):
+    """Filters the databases list down to only the ones we need, in place."""
+    dbs = rc.databases
+    public_only = rc._get("public_only", False)
+    if public_only:
+        dbs = [db for db in dbs if db["public"]]
+    dbname = rc._get("db")
+    if dbname is not None:
+        dbs = [db for db in dbs if db["name"] == dbname]
+    elif len(dbs) == 1:
+        rc.db = dbs[0]["name"]
+    rc.databases = dbs
+
+
+def connect_db(rc, colls=None):
+    '''
+    Load up the db's
+
+    Parameters
+    ----------
+    rc:
+        The runcontrol instance
+    colls
+        The list of collections that should be loaded
+
+    Returns
+    -------
+    chained_db:
+      The chained databases in the form of a document
+    dbs:
+       The databases in the form of a runcontrol client
+    '''
+    with connect(rc, dbs=colls) as rc.client:
```

Comment:
my one request here (and I would rather this was fixed in regolith tbh!) is to get the usage of dbs and collections consistent.  The way regolith works is that you can maintain different dbs (for example, a private one only you can read, a group-readable one and a public one, for example, but it can be anything...they are just physically located in different places) then there are collections (people, contacts, projects, projecta, etc.).  Regolith can load the same collection from across multiple databases, so I can have some projecta in my projecta collection in my private db and some other projecta in the group projecta db.  If I run regolith and my db list includes `private_db` and `group_db` regolith will build the collection by chaining what it finds in each db.  But if I run from a different place where the `regolithrc.json` only specifies the `group_db` the collection built will only have the prums from the projecta collection in the group_db and so on.     

This is more sophisticated than we need in pydatarecognition because we aim to just set this up on an app with a single database backend.   However, it probably makes more sense to take it as it is without modifying it (which could involve a lot of work) and just use it.